### PR TITLE
Feat: profile calls

### PR DESCRIPTION
**Pull request by bout3fiddy**
Enables call profiling. User needs to set up their profiling tests similar to the following example:

```
@pytest.mark.parametrize(
    "a,b,c", [(42, 69, 1), (420, 690, 2), (42, 690, 3), (420, 69, 4)]
)
@pytest.mark.profile_calls("boa_contract.foo", "boa_contract.bar")
def test_populate_call_profile_property(boa_contract, a, b, c):

    boa_contract.foo(a, b)
    boa_contract.bar(a, b)

    assert boa_contract.call_profile
```

which generates:
```
tests/unitary/test_call_profiling/test_call_profile.py .....                                                                                     [100%] 
                                               
                  Call Profile                  
┏━━━━━━━━┳━━━━━━┳━━━━━━━━┳━━━━━━━┳━━━━━━┳━━━━━━┓
┃ Method ┃ Mean ┃ Median ┃ Stdev ┃ Min  ┃ Max  ┃
┡━━━━━━━━╇━━━━━━╇━━━━━━━━╇━━━━━━━╇━━━━━━╇━━━━━━┩
│    foo │ 188  │ 188.0  │ 0     │ 188  │ 188  │
│    bar │ 5275 │ 5275.0 │ 0     │ 5275 │ 5275 │
└────────┴──────┴────────┴───────┴──────┴──────┘
```

### CODE REVIEW

File: boa/test/plugin.py [hunk 1]
```diff
@@ -27,13 +27,129 @@ def f(*args, **kwargs):
...
+    profile_calls = item.get_closest_marker("profile_calls").args
+    if len(profile_calls) > 0:
+
+        for call in profile_calls:
+            fixture_name = call.split(".")[0]
+            for name, fixture in item.funcargs.items():
+
+                if name == fixture_name:
+
+                    fixture.profile_calls = True
```
Comment (charles-cooper) [inline boa/test/plugin.py:57]: this has a couple problems. first of all, `profile_calls` is not unset at the end of the item. second of all, the approach is brittle. it relies on parsing the pytest arguments at runtime. what if there is a collision between fixture names and called function names?

it's going to be much cleaner if `enable_call_profiling` sets a flag globally (or on `boa.env`) which can be checked by all contracts at execution, and it's easier to reason about the state of the flag (set/unset) and has less dependency to set up and teardown, since the flag is either set or unset.

**Review by charles-cooper — COMMENTED** [review]
i see this is a draft, but i left some feedback on the approach. i think it will be much cleaner to reason about if the `profiling` flag is set in a more global place, rather than on a per-contract basis. the current approach does not clear the flag, so it has a memory leak because all tests will end up profiling (and not clearing the profiling data).

i think it might also be cleaner to set it up so that profiling happens "lazily", at the end of the entire test suite, rather than right after each test. this is going to be more performant, since the profiler path will only be taken once as a batch action at the end of the test suite, and also feel more performant (to the user), since profiling won't happen during the execution of the tests themselves, and we may even be able to disable profile report generation in a lot of cases (ex. if the test suite fails). lastly, it will be more general, since if we want to add other instrumentation later, we will not need to do it at the time of the test, we can just calculate it "after the fact". so the approach i suggest is, to just append computations an Environment-scoped variable, keeping track of all computation objects that have happened during profiling tests. then at the end of the test suite, we can iterate over them and build a profile object. there may be some undesirable edge cases which have less good reasonability properties than the "eager" way, but besides contract code changing mid-test suite, i'm having trouble thinking of good ones.

**Comment by bout3fiddy** [comment]
Following comments from @charles-cooper, I've introduced changes that allow lazily profiling calls. Usage:

```
@pytest.mark.profile_calls
def test_something(contract):
    contract.foo()
```

**Comment by bout3fiddy** [comment]
not sure how to make this table look pretty in reStructuredText:

image

File: boa/test/plugin.py [hunk 2]
```diff
@@ -15,8 +16,15 @@ def _HypothesisHandle__init__(self, *args, **kwargs):
...
+    has_profiling_flag = False
+    if "pytestmark" in t.__dict__:
+        for mark in t.pytestmark:
+            if mark.name == "profile_calls":
+                has_profiling_flag = True
+                break
+
     def f(*args, **kwargs):
-        with boa.env.anchor():
+        with boa.env.anchor(), boa.env.store_call_profile(has_profiling_flag):
```
Comment (charles-cooper) [inline boa/test/plugin.py:27]: i don't think this is required to do on every test run, since the pytest machinery should already take care of it.

File: boa/environment.py [hunk 3]
```diff
@@ -266,6 +266,9 @@ def __init__(self):
 
         self._contracts = {}
 
+        self.profile_calls = False
+        self.profiled_calls = {}
```
Comment (charles-cooper) [inline boa/environment.py:270]: both of these should be private

File: boa/environment.py [hunk 4]
```diff
@@ -341,6 +344,14 @@ def gas_meter_class(self, cls):
...
         self.set_gas_meter_class(ProfilingGasMeter)
 
+    @contextlib.contextmanager
+    def store_call_profile(self, enable: bool = True) -> None:
+        try:
+            self.profile_calls = enable
+            yield
+        finally:
+            self.profile_calls = False
+
```
Comment (charles-cooper) [inline boa/environment.py:354]: I'm not convinced that we need to expose this on the environment object. I can't think of why a user would want to use it -- it's pretty much only there for testing purposes at this time. instead of having this method, we should just inline this code in `pytest_runtest_call`

File: boa/vyper/contract.py [hunk 5]
```diff
@@ -585,6 +588,20 @@ def marshal_to_python(self, computation, vyper_typ):
...
+        if self.env.profile_calls:
+            try:
+                fn_name = self._get_fn_from_computation(computation).name
+            except AttributeError:
+                # TODO: https://github.com/vyperlang/vyper/issues/3200  # noqa: E501
+                fn_name = "unnamed"
+            fn_name = self.compiler_data.contract_name + "." + fn_name
+            gas_used = computation.get_gas_used()
+            if fn_name not in self.env.profiled_calls.keys():
+                self.env.profiled_calls[fn_name] = [gas_used]
```
Comment (charles-cooper) [inline boa/vyper/contract.py:601]: there is a problem here if two contracts have the same name and have a function with the same name.

File: pyproject.toml [hunk 6]
```diff
@@ -21,17 +21,15 @@ dependencies = [
     "eth-typing",
     "hypothesis",
     "pytest",
-#'eth-stdlib; python_version >= "3.10"',
+    "rich",
```
Comment (charles-cooper) [inline pyproject.toml:24]: should this be an extra?

File: boa/environment.py (hunk 4 above)
Comment (bout3fiddy) [reply boa/environment.py:354]: Well they could use this for their own statistics. This just offers a way to capture gas used. in boa.env.profiled_calls dict. Nothing more really.

File: boa/test/plugin.py (hunk 2 above)
Comment (bout3fiddy) [reply boa/test/plugin.py:27]: it does not do so for hypothesis (which is why it is here).

File: pyproject.toml (hunk 6 above)
Comment (bout3fiddy) [reply pyproject.toml:24]: I think it's fine. In fact, we should consider using rich in more places (because it is such an awesome module for prettifying console outputs).

It doesn't really bloat up boa, and is only imported at the end of the pytest session.

File: boa/test/plugin.py [hunk 7]
```diff
@@ -1,22 +1,40 @@
...
 _old_init = hypothesis.core.HypothesisHandle.__init__
 
 
+@contextlib.contextmanager
+def _enable_profiling(flag: bool):
+    try:
+        boa.env._profile_calls = flag
+        yield
+    finally:
+        boa.env._profile_calls = False
```
Comment (charles-cooper) [inline boa/test/plugin.py:21]: this has an interesting effect of disabling `_profile_calls` even if it was set to `True` before `_enable_profiling()` was called.

File: boa/profiling.py [hunk 8]
```diff
@@ -133,3 +137,125 @@ def summary(
...
+
+
+# simple dataclass for fn
+SelectorInfo = namedtuple("SelectorInfo", ["fn_name", "contract_name", "address"])
+
+
+# cache gas_used for all computation (including children)
+def cache_gas_used_for_computation(computation, env):
+
+    contract = env.lookup_contract(computation.msg.storage_address)
```
Comment (charles-cooper) [inline boa/profiling.py:149]: this could fail for a black box contract

Comment (charles-cooper) [reply boa/profiling.py:149]: it might be useful to use the `Datum` abstraction earlier in this file

File: boa/profiling.py [hunk 9]
```diff
@@ -133,3 +137,125 @@ def summary(
...
+        # https://github.com/vyperlang/vyper/pull/3202
+        fn_name = "unnamed"
+
+    fn = SelectorInfo(
+        fn_name=fn_name,
+        contract_name=contract.compiler_data.contract_name,
+        address=to_checksum_address(contract.address),
+    )
+
+    gas_used = computation.get_gas_used()
```
Comment (charles-cooper) [inline boa/profiling.py:164]: doesn't handle gas refund -- check `Datum` abstraction as mentioned here https://github.com/vyperlang/titanoboa/pull/48/files/4349f08f1c5964592ee1800fd3a37e5d04a597b5#r1051643690

File: boa/test/plugin.py (hunk 7 above)
Comment (bout3fiddy) [reply boa/test/plugin.py]: renamed it to `_toggle_profiling`

File: boa/vyper/contract.py [hunk 10]
```diff
@@ -465,15 +466,17 @@ def at(self, *args, **kwargs):
     def ast_map(self):
         return ast_map_of(self.compiler_data.vyper_module)
 
+    def _get_fn_from_computation(self, computation):
+        node = self.find_source_of(computation.code)
+        if isinstance(node, vy_ast.FunctionDef):
```
Comment (charles-cooper) [inline boa/vyper/contract.py:471]: i think this breaks the previous None check?

Comment (bout3fiddy) [reply boa/vyper/contract.py:471]: It will not, since that is in the profiling section https://github.com/vyperlang/titanoboa/blob/dae534980793c36704f97f53bdecc28b8bb591ee/boa/profiling.py#L181-L188

File: boa/profiling.py [hunk 11]
```diff
@@ -1,8 +1,20 @@
...
+
+from boa.environment import Env
 from boa.vyper.ast_utils import get_line
 
+# simple dataclass for fn
+CallInfo = namedtuple("SelectorInfo", ["fn_name", "contract_name", "address"])
+LineInfo = namedtuple(
+    "LineGasUsedInfo", ["address", "contract_name", "lineno", "line_src"]
+)
+
```
Comment (charles-cooper) [inline boa/profiling.py:17]: why not just use actual dataclasses here?

File: boa/test/plugin.py [hunk 12]
```diff
@@ -1,15 +1,27 @@
...
+from boa.profiling import get_call_profile_table, get_line_profile_table
+from boa.vm.gas_meters import ProfilingGasMeter
 
 # monkey patch HypothesisHandle. this fixes underlying isolation for
 # hypothesis.given() and also hypothesis stateful functionality.
 _old_init = hypothesis.core.HypothesisHandle.__init__
 
 
+@contextlib.contextmanager
+def _toggle_profiling(yes: bool = False):
```
Comment (charles-cooper) [inline boa/test/plugin.py:17]: this function is confusing. if the flag is False, it doesn't disable profiling. also, it's only used in one place - maybe just inline it there?

File: boa/profiling.py [hunk 13]
```diff
@@ -102,34 +117,267 @@ def merge(self, other: "LineProfile") -> None:
...
+        else:
+            env._cached_line_profiles[line].append(gas_used)
+
+    # ------------------------------------------------------------
+
+    # recursion for child computations
+    for _computation in computation.children:
+        child_contract = env.lookup_contract(_computation.msg.code_address)
+        # ignore black box contracts
+        if child_contract:
```
Comment (charles-cooper) [inline boa/profiling.py:216]:
```suggestion
        if child_contract is not None:
```

File: boa/profiling.py [hunk 14]
```diff
@@ -102,34 +117,267 @@ def merge(self, other: "LineProfile") -> None:
...
+        contract_uid = (lp.contract_name, lp.address)
+
+        # add spaces so numbers take up equal space
+        lineno = str(lp.lineno)
+        if lp.lineno < 10:
+            lineno = "   " + lineno
+        if 10 <= lp.lineno < 100:
+            lineno = "  " + lineno
+        if 100 <= lp.lineno < 1000:
+            lineno = " " + lineno
```
Comment (charles-cooper) [inline boa/profiling.py:324]: this whole block can just be a call to `rjust`

File: boa/profiling.py [hunk 15]
```diff
@@ -102,34 +117,267 @@ def merge(self, other: "LineProfile") -> None:
...
+
+        gas_data = (lineno + ": " + lp.line_src, gas_data)
+        if contract_uid not in contracts:
+            contracts[contract_uid] = [gas_data]
+        else:
+            contracts[contract_uid].append(gas_data)
```
Comment (charles-cooper) [inline boa/profiling.py:330]: small thing but this will be more concise with a call to `setdefault()`:
```suggestion
        s = contracts.setdefault(contract_uid, [])
        s.append(gas_data)
```

File: boa/profiling.py [hunk 16]
```diff
@@ -102,34 +117,267 @@ def merge(self, other: "LineProfile") -> None:
...
+                    len(gas_used),
+                    int(statistics.mean(gas_used)),
+                    int(statistics.median(gas_used)),
+                    stdev,
+                    min(gas_used),
+                    max(gas_used),
+                ]
+            )
+
+        l_profile = sorted(l_profile, key=lambda x: int(x[3]), reverse=True)
```
Comment (charles-cooper) [inline boa/profiling.py:359]: comment what `x[3]` is please

File: boa/profiling.py [hunk 17]
```diff
@@ -102,34 +117,267 @@ def merge(self, other: "LineProfile") -> None:
...
+                [
+                    contract_name,
+                    code,
+                    len(gas_used),
+                    int(statistics.mean(gas_used)),
+                    int(statistics.median(gas_used)),
+                    stdev,
+                    min(gas_used),
+                    max(gas_used),
+                ]
```
Comment (charles-cooper) [inline boa/profiling.py:356]: style: unless the item should be mutable, make it a tuple

Comment (charles-cooper) [reply boa/profiling.py:356]: since this code is similar to the long block in `get_call_profile_table()`, can we factor it into a Stats object?

**Review by charles-cooper — COMMENTED** [review]
looking good, some style changes requested. also there are long blocks which are just moving around clumps of stats results, can you please refactor those into a class and maybe some convenience methods for working with it? you can get `list()` behavior by implementing iterator or bound indexing on the object: https://stackoverflow.com/a/30554680
